src/ast/rewriter/seq_rewriter.cpp

```diff
@@ -2042,7 +2211,149 @@ expr_ref seq_rewriter::is_nullable(expr* r) {
     return result;
 }
 
+br_status seq_rewriter::derivative(expr* hd, expr* r, expr_ref& result) {
+    // Check assumption: hd is a single character string
+    // TODO: I want to check if is_char, but there doesn't seem to be an option for that?
+    SASSERT(m_util.str.is_unit(hd));
+    expr* r1 = nullptr;
+    expr* r2 = nullptr;
+    expr* p = nullptr;
+    // unsigned lo = 0, hi = 0;
+    if (m_util.re.is_concat(r, r1, r2)) {
+        expr_ref is_n = is_nullable(r1);
+        expr_ref dr1(m());
+        expr_ref dr2(m());
+        br_status status1 = derivative(hd, r1, dr1);
+        br_status status2 = derivative(hd, r2, dr2);
+        if (status1 != BR_DONE) {
+            return BR_FAILED;
+        }
+        else if (m().is_false(is_n)) {
+            result = m_util.re.mk_concat(dr1, r2);
+            return BR_DONE;
+        }
+        else if (status2 != BR_DONE) {
+            return BR_FAILED;
+        }
+        else if (m().is_true(is_n)) {
+            result = m_util.re.mk_union(
+                m_util.re.mk_concat(dr1, r2),
+                dr2
+            );
+            return BR_DONE;
+        }
+        else {
+            result = m_util.re.mk_union(
+                m_util.re.mk_concat(dr1, r2),
+                kleene_and(is_n, dr2)
+            );
+            return BR_DONE;
+        }
+    }
+    else if (m_util.re.is_star(r, r1)) {
+        expr_ref dr1(m());
+        if (derivative(hd, r1, dr1) == BR_DONE) {
+            result = m_util.re.mk_concat(dr1, r);
+            return BR_DONE;
+        } else {
+            return BR_FAILED;
+        }
+    }
+    else if (m_util.re.is_plus(r, r1)) {
+        return derivative(hd, m_util.re.mk_star(r1), result);
+    }
+    else if (m_util.re.is_union(r, r1, r2)) {
```

Comment:
essentially, shortcutting on union and intersection is possible. Suppose the full set is returned by one of the branches, then the result is the full set. For intersection, if one of the branches is empty, the result can be empty. The return value of the other branch is not relevant. 

expr_ref dr1 = derivative(hd, r1);
if (dr1 && m_util.re.is_full_re(dr1))
      return dr1;
expr_ref dr1 = derivative(hd, r1);
if (dr2 && m_util.re.is_full_re(dr2))
      return dr2;
if (!dr1 || !dr2)
       return expr_ref(m());
return expr_ref(m_util.re.mk_union(dr1, dr2), m());

Reply:
I didn't address this comment yet, because I think returning BR_REWRITE_FULL after each iteration of the derivative should result in simplifying such cases. This is line 2251. Should we also do the simplification in the union and intersections directly?